### ISSUE TITLE
Integrate privly-applications selenium tests

### ISSUE THREAD

**Issue by smcgregor** [privly/privly-safari#20]
The selenium tests found in privly-applications test all the core functionality of the system and will quickly tell you if there is an issue integrating privly-applications with privly-safari. These tests should be supported by the Safari browser's extension context and/or by the Safari web (hosted) context.

Steps:

1. Install the [Safari Driver](https://code.google.com/p/selenium/wiki/SafariDriver)
1. Add safari_web and safari_extension to the [platforms list](https://github.com/privly/privly-applications/blob/master/test/selenium/run_all.rb#L20) along with [details](https://github.com/privly/privly-applications/blob/master/test/selenium/run_all.rb#L90).
2. Add the platforms to [this list](https://github.com/privly/privly-applications/blob/master/test/selenium/run_all.rb#L145) along with the proper sanity check.
3. Add the corresponding configurations for [safari_web](https://github.com/privly/privly-applications/blob/master/test/selenium/configurations.rb#L122) and [safari_extension](https://github.com/privly/privly-applications/blob/master/test/selenium/configurations.rb#L187).
4. [Run the tests](https://github.com/privly/privly-applications/blob/master/test/selenium/run_all.rb) with `ruby run_all.rb --platform safari_web --release-status experimental --content-server http://localhost:3000` For the tests to pass you will need a local content server running on port 3000. You do not need the tests to pass to see that you have configured testing correctly.

I recommend getting this working for safari_web before attempting to test safari_extension.

Completing these steps for TravisCI and SauceLabs will likely be more complicated since it would require somehow installing the safari extension on SauceLabs. This works for the Firefox extension by adding it to the browser's profile from Travis, and it works for the Chrome extension by base64 encoding the extension and streaming it to SauceLabs. One of these approaches will likely work for Safari as well, but the first step is to get these tests running locally without SauceLabs and TravisCI.

**Comment by sammyshj** [comment]
@smcgregor Presently, the safari tests run locally and on SauceLabs without the extension.
The next thing I am targeting is to run the selenium tests locally with the extension. I will have to implement a function `configure_for_safari_extension` but I am not completely sure how to go about it. Can you link me to some documentation or some existing examples which help me do the same?
Thanks

**Comment by sammyshj** [comment]
@smcgregor The PR regarding running the tests locally on Safari with the extension is https://github.com/privly/privly-applications/pull/260

**Comment by smcgregor** [comment]
I think Apple may have removed the ability to programmatically install extensions via Selenium for security reasons. I recommend starting with locally testing the browser extension by pre-installing it to Safari. This will mean you can test the extension from Selenium without installing it from Selenium.

Testing on SauceLabs will be trickier, but let's get local testing running before we starting thinking about that.

**Comment by sammyshj** [comment]
@smcgregor I have found https://code.google.com/p/selenium/wiki/RubyBindings#Safari . Though I am not completely sure if this will work, I am currently trying to make this work.
I will also try as you mentioned, to first install the extension and then run the Selenium tests.

**Comment by sammyshj** [comment]
@smcgregor Presently, when I install the extension in Safari, I need to know the base URL of the safari extension in order for the tests to run. But, we do not currently get the first run page to know the URL of the extension. What do you think about this?

**Comment by smcgregor** [comment]
The problem on Chrome is that the URL can cryptographically change depending on whose system they are running their development version on. Is this the case on Safari? Why can't you hard-code the address?

**Comment by sammyshj** [comment]
@smcgregor The safari extension have a URL such as `safari-extension://com.privly.privly-34B3M269R6/32c12845/`. The last 8 characters are alpha-numeric characters which change every time the extension is loaded. Hence, the extension URL cannot be hard-coded.
One way to get the extension URL without going through the first-run page, is to send a message to the extension using the JS console and getting the extension base URL as a response. This way, once we know the extension base URL, we can then set that URL as the base URL for all the selenium tests.
What do you think about this?

**Comment by smcgregor** [comment]
Ah, ok. I think you should create a version of [this file](https://github.com/privly/privly-applications/blob/master/test/selenium/tc_chrome_helper.rb). You will need to add a first-run page for it to work. It basically waits for an extension context page to open then grabs the URL from the page.

**Comment by sammyshj** [comment]
@smcgregor Yes, I have tried creating this file but the first-run page on Safari does not load up for tests as the extension is already previously installed.
I am planning to send a message from the test to the extension and getting the extension base URL as the response.

**Comment by smcgregor** [comment]
Good point. Another option would be to initiate the posting process and to grab the resulting window's URL.

**Comment by sammyshj** [comment]
@smcgregor I will open a new PR in the privly-safari repo to support sending a message to the extension and getting back the extension base URL as a response.

**Comment by sammyshj** [comment]
Getting the URL from the posting process is an option but getting the URL from the extension sendMessage itself is an easier and a safer option.

**Comment by sammyshj** [comment]
@smcgregor Can you explain how to initiate the posting process presently?

**Comment by smcgregor** [comment]
You could do this when you add the posting button to safari by clicking the
button when it appears on the website. For now, I recommend either hard
coding the url or setting it as an environmental variable.

**Comment by sammyshj** [comment]
@smcgregor I have now got the extension URL from the injected application of privly's test page. The commit can be seen at https://github.com/sammyshj/privly-applications/commit/dc8e886ecaef8663293f1f487523f492830685d6
Let me know what you think about this.
